File: src/main/java/gargoyle/xenox/util/convert/Convert.java
package gargoyle.xenox.util.convert;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public final class Convert {
    private static final Map<Class<?>, Function<String, ?>> converters = new HashMap<>();

    static {
        converters.put(String.class, (Function<String, String>) s -> s);
        converters.put(Integer.class, (Function<String, Integer>) Integer::valueOf);
        converters.put(Boolean.class, (Function<String, Boolean>) Boolean::valueOf);
        converters.put(Double.class, (Function<String, Double>) Double::valueOf);
        converters.put(Long.class, (Function<String, Long>) Long::valueOf);
    }

    private Convert() {
    }

    @SuppressWarnings("unchecked")
    public static <T> T convert(String p, Class<T> type, T defValue) {
        if (p == null) {
            return null;
        }
        if (converters.containsKey(type)) {
            return (T) converters.get(type).apply(p);
        } else {
            return defValue;
        }
    }
}

File: src/main/kotlin/gargoyle/xenox/util/convert/Convert.kt
package gargoyle.xenox.util.convert


object Convert {
    private val converters: MutableMap<Class<*>, (String)-> Any> = HashMap()

    init {
        converters[String::class.java] =  { it }
        converters[Int::class.java] = { it.toInt() }
        converters[Boolean::class.java] = { it.toBoolean() }
        converters[Double::class.java] = { it.toDouble() }
        converters[Long::class.java] = { it.toLong() }
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> convert(p: String?, type: Class<T>, defValue: T): T? =
        when {
            p == null -> null
            type in converters ->
            converters[type]!!(p) as T
            else -> defValue
        }
}
